packages/arb-bridge-eth/contracts/rollup/Rollup.sol

```diff
@@ -0,0 +1,699 @@
+     * @param losingStaker Address of the losing staker
+     */
+    function completeChallenge(address winningStaker, address losingStaker) external override {
+        // Only the challenge contract can declare winners and losers
+        require(msg.sender == inChallenge(winningStaker, losingStaker), "WRONG_SENDER");
+
+        uint256 loserStake = amountStaked(losingStaker);
+        uint256 winnerStake = amountStaked(winningStaker);
+        if (loserStake > winnerStake) {
+            loserStake -= reduceStakeTo(losingStaker, winnerStake);
+        }
+
+        uint256 amountWon = loserStake / 2;
+        increaseStakeBy(winningStaker, amountWon);
+        loserStake -= amountWon;
+        clearChallenge(winningStaker);
+
+        // TODO: deposit extra loserStake into ArbOS
+        turnIntoZombie(losingStaker);
+    }
+
+    /**
+     * @notice Remove the given zombie from nodes it is staked on, moving backwords from the latest node it is staked on
+     * @param zombieNum Index of the zombie to remove
+     * @param maxNodes Maximum number of nodes to remove the zombie from (to limit the cost of this transaction)
+     */
+    function removeZombie(uint256 zombieNum, uint256 maxNodes) external whenNotPaused {
+        require(zombieNum <= zombieCount(), "NO_SUCH_ZOMBIE");
+        address zombieStakerAddress = zombieAddress(zombieNum);
+        uint256 latestStakedNode = zombieLatestStakedNode(zombieNum);
+        uint256 nodesRemoved = 0;
+        uint256 firstUnresolved = firstUnresolvedNode();
+        while (latestStakedNode > firstUnresolved && nodesRemoved < maxNodes) {
+            INode node = getNode(latestStakedNode);
+            node.removeStaker(zombieStakerAddress);
+            latestStakedNode = node.prev();
+            nodesRemoved++;
+        }
+        if (latestStakedNode < firstUnresolved) {
+            removeZombie(zombieNum);
+        } else {
+            zombieUpdateLatestStakedNode(zombieNum, latestStakedNode);
+        }
+    }
+
+    /**
+     * @notice Remove any zombies whose latest stake is earlier than the first unresolved node
+     * @param startIndex Index in the zombie list to start removing zombies from (to limit the cost of this transaction)
+     */
+    function removeOldZombies(uint256 startIndex) public {
+        uint256 currentZombieCount = zombieCount();
+        uint256 firstUnresolved = firstUnresolvedNode();
+        for (uint256 i = startIndex; i < currentZombieCount; i++) {
+            while (zombieLatestStakedNode(i) < firstUnresolved) {
+                removeZombie(i);
+                currentZombieCount--;
+                if (i >= currentZombieCount) {
+                    return;
+                }
+            }
+        }
+    }
+
+    /**
+     * @notice Calculate the current amount of funds required to place a new stake in the rollup
+     * @return The current minimum stake requirement
+     */
+    function currentRequiredStake() public view returns (uint256) {
```

Comment:
This doesn't quite do the right thing. We want the stake to escalate if the chain could make progress but isn't.  So we need to not escalate when there are no unresolved nodes beyond latestConfirmed.  And we should start escalating after some unconfirmed node has had its deadline passed.

So I think the right approach is to say that if there are no unresolved nodes after latestConfirmed, then don't escalate.  If there are unresolved nodes, then escalate based on the time since the deadline of firstUnresolvedNode.

Reply:
I think that makes sense and it's a pretty simple change. This has been updated